mw_validate_json.go

```diff
@@ -0,0 +1,81 @@
+package main
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/xeipuuv/gojsonschema"
+
+	"github.com/TykTechnologies/tyk/apidef"
+)
+
+type ValidateJSON struct {
+	BaseMiddleware
+}
+
+func (k *ValidateJSON) Name() string {
+	return "ValidateJSON"
+}
+
+func (k *ValidateJSON) EnabledForSpec() bool {
+	for _, v := range k.Spec.VersionData.Versions {
+		if len(v.ExtendedPaths.ValidateJSON) > 0 {
+			return true
+		}
+	}
+
+	return false
+}
+
+// ProcessRequest will run any checks on the request on the way through the system, return an error to have the chain fail
+func (k *ValidateJSON) ProcessRequest(w http.ResponseWriter, r *http.Request, _ interface{}) (error, int) {
+
+	_, versionPaths, _, _ := k.Spec.Version(r)
+	found, meta := k.Spec.CheckSpecMatchesStatus(r, versionPaths, ValidateJSONRequest)
+	if !found {
+		return nil, 200
+	}
+
+	mmeta := meta.(*apidef.ValidatePathMeta)
+	if mmeta.ValidateWith64 == "" {
+		return errors.New("no schemas to validate against"), 400
+	}
+
+	rCopy := copyRequest(r)
+	bodyBytes, err := ioutil.ReadAll(rCopy.Body)
+	if err != nil {
+		return err, 400
+	}
+	defer rCopy.Body.Close()
+
+	schemaBytes, err := base64.StdEncoding.DecodeString(mmeta.ValidateWith64)
```

Comment:
might worth to cache `schemaBytes` as this might be too expensive to decode the same bas64 encoded string with every incoming through this middleware request

Reply:
@dencoded nice - suggestions on implementation / examples of where done before please?

Reply:
I feel like decoding/loading json-schema is an operation which needs to be done when we load/reload API spec definition, so we can put loaded/ready to use json-schema in some additional field and use it